includes/User.php

```diff
@@ -483,18 +494,63 @@ public static function newFromConfirmationCode( $code ) {
 	}
 
 	/**
-	 * Create a new user object using data from session or cookies. If the
-	 * login credentials are invalid, the result is an anonymous user.
+	 * Creates a MediaWiki User object based on the token given in the HTTP request.
 	 *
-	 * @param $request WebRequest object to use; $wgRequest will be used if
-	 *        ommited.
-	 * @return User object
+	 * @param  WebRequest $request The HTTP request data as an object
+	 *
+	 * @return User                A logged in User object on successful authentication,
+	 *                             or an anonymous User object on failure
 	 */
-	public static function newFromSession( WebRequest $request = null ) {
-		$user = new User;
-		$user->mFrom = 'session';
-		$user->mRequest = $request;
-		return $user;
+	public static function newFromToken( \WebRequest $request ) {
+		global $wgMemc;
+
+		$cookieHelper = self::authCookieHelper();
+		$token = $cookieHelper->getAccessToken( $request );
+
+		if ( !$token ) {
+			return new User;
+		}
+
+		try {
+			$tokenInfo = self::heliosClient()->info( $token );
+			if ( empty( $tokenInfo->user_id ) ) {
+				return new User;
+			}
+			$user = self::newFromId( $tokenInfo->user_id );
+			$user->setGlobalAuthToken( $token );
+
+			// Don't return the user object if it's disabled
+			// @see SERVICES-459
+			if ( (bool)$user->getGlobalFlag( 'disabled' ) ) {
+				$cookieHelper->clearAuthenticationCookie( $request->response() );
+				$user->loadDefaults();
+				return $user;
+			}
+
+			// start the session if there's none so far
+			// the code is borrowed from SpecialUserlogin
+			// @see PLATFORM-1261
+			if ( session_id() == '' ) {
+				$sessionId = substr( hash( 'sha256', $token ), 0, 32 );
+				wfSetupSession( $sessionId );
+				WikiaLogger::instance()->debug( __METHOD__ . '::startSession' );
+
+				// Update mTouched on user when he starts new MW session, but not too often
+				// @see SOC-1326 and SUS-546
+				$throttleKey = wfSharedMemcKey( self::INVALIDATE_CACHE_THROTTLE_SESSION_KEY, $tokenInfo->user_id );
+				$invalidateCacheThrottleTime = $wgMemc->get( $throttleKey );
+				if ( $invalidateCacheThrottleTime === null || $invalidateCacheThrottleTime < time() ) {
+					$wgMemc->set( $throttleKey, time() + self::INVALIDATE_CACHE_THROTTLE, self::INVALIDATE_CACHE_THROTTLE );
+					$user->invalidateCache();
+				}
+			}
+
+			return $user;
+		} catch ( ClientException $e ) {
+			WikiaLogger::instance()->error( __METHOD__, [ 'exception' => $e ] );
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "exception" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1192)